File: src/python/ejercicios/Ejercicio6.py
import threading
import time

recurso = threading.Lock()
accesos = {}

def hilo_rapido(id):
    for i in range(20):
        with recurso:
            accesos[id] = accesos.get(id, 0) + 1
            print(f"Hilo rapido {id} accedio (total: {accesos[id]})")

def hilo_lento():
    for i in range(5):
        with recurso:
            print(f">>> Hilo LENTO accedio (vez {i + 1})")
        time.sleep(0.05)

# Hilos rapidos
for i in range(1, 4):
    t = threading.Thread(target=hilo_rapido, args=(i,))
    t.daemon = True
    t.start()

# Hilo lento
lento = threading.Thread(target=hilo_lento)
lento.start()
lento.join()

print("\nLos hilos rapidos acaparan el recurso y el lento")
print("apenas logra acceder (starvation).")

"""
Preguntas:
1. ¿Qué significa que un hilo sufra starvation?
   Significa que nunca (o muy raramente) obtiene acceso al recurso
   porque otros hilos lo acaparan constantemente.

2. ¿Es lo mismo que deadlock?
   No. En deadlock los hilos se bloquean mutuamente y ninguno avanza.
   En starvation hay hilos que avanzan mientras uno queda postergado
   indefinidamente, pero el sistema sigue funcionando.

3. ¿Cómo se puede reducir este problema?
   - Usar colas justas (fair queues) como threading.Lock con
     scheduling equitativo.
   - Aumentar la prioridad del hilo postergado.
   - Implementar envejecimiento (aging): aumentar prioridad cuanto
     mas tiempo espera el hilo.
"""

File: src/csharp/ejercicios/Ejercicio6.cs
using System;
using System.Threading;

class Ejercicio6
{
    public static void Run()
    {
        var recurso = new object();

        for (int id = 1; id <= 3; id++)
        {
            int idCapturado = id;
            new Thread(() => {
                for (int i = 0; i < 20; i++)
                {
                    lock (recurso)
                        Console.WriteLine($"Hilo rapido {idCapturado} accedio (total: {i + 1})");
                    Thread.Yield();
                }
            }).Start();
        }

        var hiloLento = new Thread(() => {
            for (int i = 0; i < 5; i++)
            {
                lock (recurso)
                    Console.WriteLine($">>> Hilo LENTO accedio (vez {i + 1})");
                Thread.Sleep(50);
            }
        });

        hiloLento.Start();
        hiloLento.Join();

        Console.WriteLine("\nLos hilos rapidos acaparan el recurso.");
        Console.WriteLine("El lento apenas logra acceder (starvation).");
    }
}

/*
Preguntas:
1. ¿Que significa starvation? Un hilo nunca (o casi nunca) accede
   al recurso porque otros lo acaparan.
2. ¿Es lo mismo que deadlock? No. En deadlock nada avanza. En
   starvation algunos avanzan, uno queda postergado.
3. ¿Como reducirlo? Colas justas, envejecimiento (aging), prioridades.
*/
